### DISCUSSION TITLE
Problem with raspi camera on Nvidia jetson nano 2gb and running detect.py

### DISCUSSION THREAD

**Discussion by domcio01** [ultralytics/yolov5#6681] [Q&A]
Hi :) so i'm trying to run yolov5 on nvidia jetson nano 2gb with raspi camera (V2.1 i guess it is IMX219) i just need to run detect.py but got problems with camera :) some users on nvidia forum recommended to add this line 
`"nvarguscamerasrc ! video/x-raw(memory:NVMM), width=(int)1920, height=(int)1080,format=(string)NV12, framerate=(fraction)30/1 ! nvvidconv ! video/x-raw, format=(string)BGRx ! videoconvert !  appsink"`
to sv2.videocapture() but in detect.py there is no videocapture and i have no idea how to add it to source(webcam) variable.

Anyone had same problem with running detect.py with --source 0 and raspi camera ?
Hope for some of your ideas ! thank you!:)

im trying to run: 

`sudo python3 detect.py --weights yolov5n6.pt --img 640 --conf 0.25  --source 0 --device 0`

but i get:
`detect: weights=['yolov5n6.pt'], source=0, data=data/coco128.yaml, imgsz=[640, 640], conf_thres=0.25, iou_thres=0.45, max_det=1000, device=0, view_img=False, save_txt=False, save_conf=False, save_crop=False, nosave=False, classes=None, agnostic_nms=False, augment=False, visualize=False, update=False, project=runs/detect, name=exp, exist_ok=False, line_thickness=3, hide_labels=False, hide_conf=False, half=False, dnn=False
YOLOv5 🚀 v6.0-207-g8efe977 torch 1.8.0 CUDA:0 (NVIDIA Tegra X1, 1972MiB)

Fusing layers... 
Model Summary: 280 layers, 3239884 parameters, 0 gradients
[ WARN:0@333.433] global /io/opencv/modules/videoio/src/cap_v4l.cpp (1910) getProperty VIDEOIO(V4L2:/dev/video0): Unable to get camera FPS
[ WARN:0@343.524] global /io/opencv/modules/videoio/src/cap_v4l.cpp (1000) tryIoctl VIDEOIO(V4L2:/dev/video0): select() timeout.
1/1: 0...  Success (inf frames 3264x2464 at 99.00 FPS)

Traceback (most recent call last):
  File "detect.py", line 257, in <module>
    main(opt)
  File "detect.py", line 252, in main
    run(**vars(opt))
  File "/usr/local/lib/python3.6/dist-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "detect.py", line 105, in run
    dataset = LoadStreams(source, img_size=imgsz, stride=stride, auto=pt)
  File "/home/dominik/yolov5/utils/datasets.py", line 326, in __init__
    s = np.stack([letterbox(x, self.img_size, stride=self.stride, auto=self.auto)[0].shape for x in self.imgs])
  File "/home/dominik/yolov5/utils/datasets.py", line 326, in <listcomp>
    s = np.stack([letterbox(x, self.img_size, stride=self.stride, auto=self.auto)[0].shape for x in self.imgs])
  File "/home/dominik/yolov5/utils/augmentations.py", line 93, in letterbox
    shape = im.shape[:2]  # current shape [height, width]
AttributeError: 'NoneType' object has no attribute 'shape'
[ WARN:1@353.706] global /io/opencv/modules/videoio/src/cap_v4l.cpp (1000) tryIoctl VIDEOIO(V4L2:/dev/video0): select() timeout.
terminate called without an active exception
Aborted
`

**Comment by glenn-jocher** [comment]
@domcio01 I'm sorry we don't have our own jetson device to test (we should though). If normal inference, i.e. `--source image.jpg` is working then it must be related to the rpi camera like you said. Maybe a community member has some ideas.

**Comment by zeleny-ondrej** [comment]
Hi, 
does your algorithm work on image or video ? What kind of OS are you running ? Im trying to run similar algorithm on Jetson but im not able to fit in into 2GB memory.

**Reply by domcio01** [reply]
i think you wont be able ... only yolov3- tiny works for me with 8fps so im using yolov3 and im trying to train my own weights (i can upload my gcolab file but i use Colab Pro on normal RAM or GPU usage is not enought)

**Reply by zeleny-ondrej** [reply]
The issues is not training, im trying to run inference, you based on what i see above somewhat managed to run it however you had some issues with camera. Did you managed to run it on a single image ?

**Reply by domcio01** [reply]
try this: 
type in terminal: 


```
cd darknet
vim t1.sh
```


write :w and :q! to save it and quit (check vim commands i dont remember it well)
in terminal:


```
sudo apt install mcedit or sudo apt-get install -y mcedit
sudo apt update
cd darknet
mcedit ./t1.sh
```



download yolov3-tiny.weights   (https://pjreddie.com/media/files/yolov3-tiny.weights) paste them into darknet folder paste this:



```
#!/bin/bash
./darknet detector demo cfg/coco.data cfg/yolov3.cfg yolov3-tiny.weights  "nvarguscamerasrc ! video/x-raw(memory:NVMM), width=1920, height=1080, format=(string)NV12, framerate=(fraction)30/1 ! nvvidconv flip-method=0  ! video/x-raw, format=(string)BGRx ! videoconvert ! video/x-raw, format=(string)BGR ! appsink"

```


save it (F2)
reboot system 
and after it starts:
in terminal:


```
cd darknet
chmod 777 ./t1.sh
./t1.sh
```

it works for me :) later you can swap cfg and coco.data and weight on yours

**Reply by domcio01** [reply]
thats for yolov3 not for yolov5 :) i think problem is size of weights and for yolo v5 you need to run sth like this 
`python3 detect.py --weights yolov5n6.pt --img 640 --conf 0.25 --source 0 --device 0 "nvarguscamerasrc ! video/x-raw(memory:NVMM), width=1920, height=1080, format=(string)NV12, framerate=(fraction)30/1 ! nvvidconv flip-method=0  ! video/x-raw, format=(string)BGRx ! videoconvert ! video/x-raw, format=(string)BGR ! appsink"`

maybe it will work or try with making sh file like i wrote for yolov3 and run this sh file with this command or others

you need to write in terminal "python3 " and without sudo i gusse ? but not sure i found somewhere it makes some errors ??not sure for 100%

**Reply by zeleny-ondrej** [reply]
YoloV3 is not an option for me. I need to figure out the YOLOv5 issue. As i said its running on CPU just fine (just long inference time). I did not use PyTorch container because on not too familiar with them. Im currently figuring out how to run it in container on python3.6.

But what im wondering if you tried running the yolov5 on a image source not camera. From what i see in you original post, you've managed to get past the model build part but got stuck on dataloader ( LoadStream). 
What kind of OS were you running ? Is it the Jetson Nano 2GB image or did you use something else ? 
How did you installed your dependencies ? (since pytorch will only support CPU when installed through pip)

**Reply by zeleny-ondrej** [reply]
So, good new, i've managed to run it on experimenta Ubuntu20 with python 3.8 and pytorch already installed. I managed to fit it in by extending the swap size to 4 GB however, it is still slow on GPU (about 6sec/frame), actually it is much faster when running on CPU (1s/frame) so im guessing there are some issue with on which device the particullar thing is (most of the preprocessing stuff is on CPU). 

In case you want to replicate my setup, here are links to what i have installed:
https://github.com/Qengineering/Jetson-Nano-Ubuntu-20-image
https://github.com/JetsonHacksNano/resizeSwapMemory
disable desktop enviroment using **sudo telinit 3**

pip install everything but torch, torchvision, torchaudio 

run inference on videos/images
python3.8 detect.py --imgsz 640 --weights etc.
